### PR TITLE
yul proto: Add support for generating string and hex literals.

### LINKED ISSUE #6307: yul proto spec: Add support for generating string and hex literals
## Abstract

It would be nice to have support for string and hex literals so that we can generate code such as 
let x := "hello"
let y := 0x1234

This issue tracks support for this.

## Motivation

- Extend current proto specification for the said literals
- Keep in mind that string and hex literals are bounded
  - Max hex character length is 64
  - Max string literal length is 32

For the latter, see

https://github.com/ethereum/solidity/blob/834c23fc38f4212a30018cab0414db9323886f5d/test/libsolidity/InlineAssembly.cpp#L658-L661

## Specification

The technical spec. mirrors [yul grammar](https://solidity.readthedocs.io/en/v0.5.6/yul.html)

### PR DESCRIPTION
**Pull request by bshastry**
### Description

(closes #6307 )

This PR adds support for generating string and hex literals in fuzzing input for the proto-based differential fuzzer.

### CODE REVIEW

**Comment by bshastry** [comment]
Chris comments that createHex and createAlphaNum may return string instead of const string.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 1]
```diff
@@ -23,13 +23,51 @@
 using namespace std;
 using namespace yul::test::yul_fuzzer;
 
+std::string const yul::test::yul_fuzzer::createHex(std::string const& _hexBytes)
```
Comment (axic) [inline test/tools/ossfuzz/protoToYul.cpp]: There's `using namespace std` above, no need to prefix these.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 2]
```diff
@@ -23,13 +23,51 @@
...
+		}), tmp.end());
+		tmp = tmp.substr(0, 64);
+	}
+	// We need this awkward if case because tmp can also be null post erase call above
+	if (tmp.empty())
+		tmp = "1";
+	return tmp;
+}
+
+std::string const yul::test::yul_fuzzer::createAlphaNum(std::string const& _strBytes)
```
Comment (axic) [inline test/tools/ossfuzz/protoToYul.cpp]: Please remove the `const` so it will be moved properly.

File: test/tools/ossfuzz/protoToYul.cpp [hunk 3]
```diff
@@ -23,13 +23,51 @@
...
 	switch (_x.literal_oneof_case())
 	{
 		case Literal::kIntval:
 			_os << _x.intval();
 			break;
+		case Literal::kHexval:
+			_os << "0x" << createHex(_x.hexval());
+			break;
+		case Literal::kStrval:
+			_os << "\"" << createAlphaNum(_x.strval()) << "\"";
```
Comment (axic) [inline test/tools/ossfuzz/protoToYul.cpp:67]: Do we really have support for string literals?

File: test/tools/ossfuzz/protoToYul.cpp (hunk 1 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: Yeah, this was suggested before. There is a problem with the compiler toolchain that the fuzzer relies on. tl;dr sadly we need to keep the `std::` prefix here.

File: test/tools/ossfuzz/protoToYul.cpp (hunk 2 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: ack

File: test/tools/ossfuzz/protoToYul.cpp (hunk 3 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp:67]: This works :)

```
$ echo {let x := add("hello", "world")} | ./solc/solc --strict-assembly -
```

iiuc, string literals are another way to encode literal values

File: test/tools/ossfuzz/protoToYul.cpp (hunk 1 above)
Comment (axic) [reply test/tools/ossfuzz/protoToYul.cpp]: Do you mean that `string` also exists in another imported namespace in the case of the fuzzer toolchain?

File: test/tools/ossfuzz/protoToYul.cpp [hunk 4]
```diff
@@ -23,13 +23,51 @@
...
+{
+	std::string tmp{_strBytes};
+	if (!tmp.empty())
+	{
+		tmp.erase(std::remove_if(tmp.begin(), tmp.end(), [](char c) {
+			return !(std::isalpha(c) || std::isdigit(c));
+		}), tmp.end());
+		tmp = tmp.substr(0, 32);
+	}
+	// We need this awkward if case because tmp can also be null post erase call above
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Actually - wouldn't the empty string be a valid literal?

File: test/tools/ossfuzz/protoToYul.cpp [hunk 5]
```diff
@@ -23,13 +23,51 @@
 using namespace std;
 using namespace yul::test::yul_fuzzer;
 
+std::string yul::test::yul_fuzzer::createHex(std::string const& _hexBytes)
+{
+	std::string tmp{_hexBytes};
+	if (!tmp.empty())
+	{
+		tmp.erase(std::remove_if(tmp.begin(), tmp.end(), [](char c) {
```
Comment (chriseth) [inline test/tools/ossfuzz/protoToYul.cpp]: Please use https://www.boost.org/doc/libs/1_68_0/libs/range/doc/html/range/reference/algorithms/new/remove_erase_if.html

File: test/tools/ossfuzz/protoToYul.cpp (hunk 1 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: @axic Yes, the fuzzing binary is linked against libc++ and the `std::` here ensures we link against the legacy (libstdc++) implementation and not libc++.

I don't know why we have competing implementations of the standard library though :man_shrugging:

File: test/tools/ossfuzz/protoToYul.cpp (hunk 4 above)
Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: Returning empty string increases the number of parsing errors because code like this is invalid

```
switch x
case "" {}
```

Comment (chriseth) [reply test/tools/ossfuzz/protoToYul.cpp]: Perhaps we should change the parser in this case? If we allow string literals, then the empty string literal should also be allowed as case label.

Comment (bshastry) [reply test/tools/ossfuzz/protoToYul.cpp]: I'm wrong, the problem is not with empty string literals but with empty hex literals. These are parsed successfully.

```
{
        let x := 1
        if ""
        {
        }
}
```

and

```
{
        let x := 1
        switch x                           
        case "" {
        }
}
```

The problem is with the following.

```
$ echo "{let x := 1 switch x case 0x {}}" | ./solc/solc - --strict-assembly
<stdin>:1:27: Error: Literal, identifier or instruction expected.
{let x := 1 switch x case 0x {}}
                          ^^
```

tl;dr PR should permit empty string literals :)

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
I think this looks good and the questions from @chriseth were addressed.
